### PR TITLE
Focus order

### PR DESCRIPTION
**Pull request by darrenburns**
Updates focus order to work in left to right then top to bottom order.
Anchor point is top left corner of the widget region **including** the margin spacing of the widget.

### CODE REVIEW

File: src/textual/dom.py [hunk 1]
```diff
@@ -596,3 +600,8 @@ def has_pseudo_class(self, *class_names: str) -> bool:
 
     def refresh(self, *, repaint: bool = True, layout: bool = False) -> None:
         pass
+
+
+def _focus_sort_key(widget: Widget) -> tuple[int, int]:
```
Comment (willmcgugan) [inline src/textual/dom.py:605]: I have a feeling we can optimize this with a combination of `attrgetter` and a new property.

It's not crazy to imagine a layout with 1000s of widgets.

I'm thinking that widget could grow a `_focus_sort_key` attribute. which return a tuple of (y, x), and we can get that with `attrgetter`

File: src/textual/widget.py [hunk 2]
```diff
@@ -494,6 +494,16 @@ def window_region(self) -> Region:
         window_region = self.region.at_offset(self.scroll_offset)
         return window_region
 
+    @property
+    def region_with_margin(self) -> Region:
```
Comment (willmcgugan) [inline src/textual/widget.py:498]: Maybe this should be named `virtual_region_with_margin`. I've used the undecorated term "region" to mean the region relative to the screen.

**Review by willmcgugan — CHANGES_REQUESTED** [review]
Nice simple solution.

Just a request which I think will cut a few cpu cycle.

File: src/textual/widget.py (hunk 2 above)
Comment (darrenburns) [reply src/textual/widget.py]: Updated

File: src/textual/dom.py (hunk 1 above)
Comment (darrenburns) [reply src/textual/dom.py]: I've made this change, and also moved `focusable_children` down to `Widget`. It's only used in the context of `Widget` types elsewhere, and in calculating it requires knowing the `virtual_region`, which doesn't exist at the `DOMNode` level.

**Review by willmcgugan — APPROVED** [review]

**Pull request merged**